File: zig/binary-search/binary_search.zig
pub fn binarySearch(T: type, target: T, items: []const T) ?usize {
    var left: usize = 0;
    var right = items.len;
    while (left < right) {
        const mid = (left + right) / 2;
        if (items[mid] < target) {
            left = mid + 1;
        } else if (items[mid] > target) {
            right = mid;
        } else {
            return mid;
        }
    }
    return null;
}

File: c/binary-search/binary_search.c
#include "binary_search.h"

const int *binary_search(int value, const int *arr, size_t length) {
  size_t l = 0, r = length;
  while (l < r) {
    const size_t m = (l + r) / 2;
    if (arr[m] < value) {
      l = m + 1;
    } else if (arr[m] > value) {
      r = m;
    } else {
      return arr + m;
    }
  }
  return NULL;
}
